src/library/scala/collection/mutable/OrderAwareHashMap.scala

```diff
@@ -0,0 +1,800 @@
+              return true
+          }
+        }
+        false
+      }
+    }
+
+    def next(): R =
+      if(!hasNext) Iterator.empty.next()
+      else node match {
+        case n: RBNode =>
+          val r = extract(n)
+          node = OrderAwareHashMap.successor(n )
+          r
+        case n: LLNode =>
+          val r = extract(n)
+          node = n.next
+          r
+      }
+  }
+
+  override def keysIterator: Iterator[K] = new MapIterator[K] {
+    protected[this] def extract(node: LLNode) = node.key
+    protected[this] def extract(node: RBNode) = node.key
+  }
+
+  override def iterator: Iterator[(K, V)] = new MapIterator[(K, V)] {
+    protected[this] def extract(node: LLNode) = (node.key, node.value)
+    protected[this] def extract(node: RBNode) = (node.key, node.value)
+  }
+
+  private[this] def growTable(newlen: Int) = {
+    var oldlen = table.length
+    table = java.util.Arrays.copyOf(table, newlen)
+    threshold = newThreshold(table.length)
+    while(oldlen < newlen) {
+      var i = 0
+      while (i < oldlen) {
+        val old = table(i)
+        if(old ne null) splitBucket(old, i, i + oldlen, oldlen)
+        i += 1
+      }
+      oldlen *= 2
+    }
+  }
+
+  @`inline` private[this] def reallocTable(newlen: Int) = {
+    table = new Array(newlen)
+    threshold = newThreshold(table.length)
+  }
+
+  @`inline` private[this] def splitBucket(tree: Node, lowBucket: Int, highBucket: Int, mask: Int): Unit = tree match {
+    case t: LLNode => splitBucket(t, lowBucket, highBucket, mask)
+    case t: RBNode => splitBucket(t, lowBucket, highBucket, mask)
+  }
+
+  private[this] def splitBucket(tree: LLNode, lowBucket: Int, highBucket: Int, mask: Int): Unit = {
+    val preLow: LLNode = new LLNode(null.asInstanceOf[K], 0, null.asInstanceOf[V], null)
+    val preHigh: LLNode = new LLNode(null.asInstanceOf[K], 0, null.asInstanceOf[V], null)
+    //preLow.next = null
+    //preHigh.next = null
+    var lastLow: LLNode = preLow
+    var lastHigh: LLNode = preHigh
+    var n = tree
+    while(n ne null) {
+      val next = n.next
+      if((n.hash & mask) == 0) { // keep low
+        lastLow.next = n
+        lastLow = n
+      } else { // move to high
+        lastHigh.next = n
+        lastHigh = n
+      }
+      n = next
+    }
+    lastLow.next = null
+    if(tree ne preLow.next) table(lowBucket) = preLow.next
+    if(preHigh.next ne null) {
+      table(highBucket) = preHigh.next
+      lastHigh.next = null
+    }
+  }
+
+  private[this] def splitBucket(tree: RBNode, lowBucket: Int, highBucket: Int, mask: Int): Unit = {
+    var lowCount, highCount = 0
+    tree.foreachNode((n: RBNode) => if((n.hash & mask) != 0) highCount += 1 else lowCount += 1)
+    if(highCount != 0) {
+      if(lowCount == 0) {
+        table(lowBucket) = null
+        table(highBucket) = tree
+      } else {
+        table(lowBucket) = fromNodes(new OrderAwareHashMap.RBNodesIterator(tree).filter(n => (n.hash & mask) == 0), lowCount)
```

Comment:
Is it worth at this point, checking if the lowCount/highCounts are below the RB tree threshold and switching back to a linked list in that case?

Reply:
The main reason for treeification is to guard against hash collision attacks. It is unlikely to occur by accident and should be quite rare in practice. And unlike java.util.HashMap, this map (just like s.c.m.HashMap/HashSet) does not support any form of shrinking, so it didn't seem worthwhile to implement untreeification.